indico/modules/core/client/js/session_expiration.js

```diff
@@ -0,0 +1,201 @@
+    }
+  }, [operation, checkTimeToExpire]);
+
+  const refreshSession = async () => {
+    setInProgress(true);
+    let response;
+    try {
+      response = await indicoAxios.post(sessionRefreshURL());
+    } catch (e) {
+      return handleAxiosError(e);
+    }
+    const freshSessionExpiration = response.data.session_expiration;
+    if (!freshSessionExpiration) {
+      dispatch({type: 'RESET'});
+      return false;
+    }
+    localStorage.setItem('session-expiration', freshSessionExpiration);
+    setInProgress(false);
+    opened.current = false;
+    clearInterval(countdownInterval.current);
+    dispatch({type: 'ACTIVATE_SESSION_TIMER'});
+    return true;
+  };
+
+  const loginAgain = () => {
+    dispatch({type: 'RESET'});
+    const hash = location.hash;
+    location.href = loginURL({next: location.href + hash});
+    return true;
+  };
+
+  return (
+    <Modal
+      size="tiny"
+      closeIcon={false}
+      closeOnDimmerClick={!inProgress}
+      closeOnEscape={!inProgress}
+      open={operation === 'countdown' || operation === 'expired'}
+    >
+      <Modal.Header>
+        {operation === 'countdown' && (
+          <Translate>Your current session is about to expire.</Translate>
+        )}
+        {operation === 'expired' && <Translate>This session has expired.</Translate>}
+      </Modal.Header>
+      <Modal.Content>
+        {operation === 'countdown' && (
+          <>
+            <Translate>
+              Click on the refresh button if you wish to continue using Indico or you will be logged
+              out.
+            </Translate>
+            <center className="ui large header">{count}</center>
+          </>
+        )}
+        {operation === 'expired' && (
+          <Translate>Your current session has expired and you have been logged out.</Translate>
+        )}
+      </Modal.Content>
+      <Modal.Actions>
+        {operation === 'countdown' && (
+          <Button
+            onClick={hardExpiry ? loginAgain : refreshSession}
+            disabled={inProgress}
+            loading={inProgress}
+          >
+            <Translate>Refresh</Translate>
+          </Button>
+        )}
+        {operation === 'expired' && (
+          <Button onClick={loginAgain} disabled={inProgress} loading={inProgress}>
+            <Translate>Login again</Translate>
+          </Button>
+        )}
+        <Button onClick={() => dispatch({type: 'RESET'})} disabled={inProgress}>
+          <Translate>Cancel</Translate>
+        </Button>
+      </Modal.Actions>
+    </Modal>
+  );
+}
+
+SessionExpirationManager.propTypes = {
+  hardExpiry: PropTypes.bool.isRequired,
+};
+
+document.addEventListener('DOMContentLoaded', () => {
+  const root = document.querySelector('#session-expiration-root');
+  if (root) {
+    const sessionExpiration = root.dataset.sessionExpiration;
+    const hardExpiry = root.dataset.hardExpiry !== undefined;
+    localStorage.setItem('session-expiration', sessionExpiration);
```

Comment:
Why do we need it in localStorage?

Reply:
This is because power users would usually have multiple tabs open at the same time. Having it in the local storage helps to synchronise the countdown across all the open tabs. Otherwise then there'll be a situation where the session has been refreshed on one tab but the other open tabs would still continue to countdown. Basically users would find this annoying

Reply:
Nice catch! :)

https://developer.mozilla.org/en-US/docs/Web/API/Broadcast_Channel_API could be an interesting option for this

Reply:
I've changed the code to make use of broadcast channels for synchronising the session expiry countdown across many tabs.